### PR TITLE
Using zipped file in create/edit data source and oracle script message

### LINKED ISSUE #1525: Using zipped file in create/edit data source and oracle script message
To reduce the size of transaction. Our chain can receive zipped file of script and compiled wasm script in message, and extract it to save in store

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #1525

### CODE REVIEW

File: chain/pkg/gzip/gzip.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"io"
+	"io/ioutil"
+)
+
+// magic bytes to identify gzip.
+// See https://www.ietf.org/rfc/rfc1952.txt
+// and https://github.com/golang/go/blob/master/src/net/http/sniff.go#L186
+var gzipIdent = []byte("\x1F\x8B\x08")
+
+func IsGzip(src []byte) bool {
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go]: Add comment.

File: chain/pkg/gzip/gzip.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	return bytes.Equal(gzipIdent, src[0:3])
+}
+
+// uncompress returns gzip uncompressed content or given src when not gzip.
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go]: update comment to match function name. Also the comment is not accurate.

File: chain/pkg/gzip/gzip.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+func Uncompress(src []byte, maxSize int64) ([]byte, error) {
+	zr, err := gzip.NewReader(bytes.NewReader(src))
+	if err != nil {
+		return nil, err
+	}
+	zr.Multistream(false)
+
+	uncompressFile, err := ioutil.ReadAll(io.LimitReader(zr, maxSize+1))
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go]: uncompressed

File: chain/pkg/gzip/gzip.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if len(uncompressFile) > int(maxSize) {
+		return nil, errors.New("uncompress file exeed maxsize")
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go]: typo

File: chain/pkg/gzip/gzip.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package oracle
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go]: This is not correct

File: chain/pkg/gzip/gzip_test.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	zw.Write(file1)
+	zw.Close()
+	gzipFile := buf.Bytes()
+
+	accFile, err := Uncompress(gzipFile, 10)
+	require.Nil(t, err)
+	require.Equal(t, file1, accFile)
+
+	accFile, err = Uncompress(gzipFile, 2)
+	require.NotNil(t, err)
```
Comment (sorawit) [inline chain/pkg/gzip/gzip_test.go]: require.Error

File: chain/pkg/gzip/gzip_test.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+func TestUncompress(t *testing.T) {
+	file1 := []byte("file")
+	var buf bytes.Buffer
+	zw := gzip.NewWriter(&buf)
+	zw.Write(file1)
+	zw.Close()
+	gzipFile := buf.Bytes()
+
+	accFile, err := Uncompress(gzipFile, 10)
+	require.Nil(t, err)
```
Comment (sorawit) [inline chain/pkg/gzip/gzip_test.go]: require.NotError

File: chain/x/oracle/handler.go [hunk 8]
```diff
@@ -37,8 +38,18 @@ func NewHandler(k Keeper) sdk.Handler {
 }
 
 func handleMsgCreateDataSource(ctx sdk.Context, k Keeper, m MsgCreateDataSource) (*sdk.Result, error) {
+	var executable []byte
```
Comment (sorawit) [inline chain/x/oracle/handler.go]: I think we can just modify m.Executable directly.

File: chain/x/oracle/handler.go [hunk 9]
```diff
@@ -37,8 +38,18 @@ func NewHandler(k Keeper) sdk.Handler {
...
+	var err error
+	if gzip.IsGzip(m.Executable) {
+		executable, err = gzip.Uncompress(m.Executable, types.MaxExecutableSize)
+		if err != nil {
+			return nil, err
```
Comment (sorawit) [inline chain/x/oracle/handler.go]: Please wrap the error with our custom error code

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/pkg/gzip/gzip_test.go (hunk 7 above)
Comment (Benzbeeb) [reply chain/pkg/gzip/gzip_test.go:20]: require don't have NotError

Comment (Benzbeeb) [reply chain/pkg/gzip/gzip_test.go]: require doesn't have NotError

File: chain/pkg/gzip/gzip.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+		return nil, err
+	}
+	zr.Multistream(false)
+
+	uncompressed, err := ioutil.ReadAll(io.LimitReader(zr, maxSize+1))
+	if err != nil {
+		return nil, err
+	}
+	if len(uncompressed) > int(maxSize) {
+		return uncompressed, errors.New("uncompressed file exceed maxSize")
```
Comment (taobun) [inline chain/pkg/gzip/gzip.go:35]: Why not return nil?

File: chain/pkg/gzip/gzip.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+// See https://www.ietf.org/rfc/rfc1952.txt
+// and https://github.com/golang/go/blob/master/src/net/http/sniff.go#L186
+var gzipIdent = []byte("\x1F\x8B\x08")
+
+// IsGzip returns true if the input file is gzipped file.
+func IsGzipped(src []byte) bool {
+	return bytes.Equal(gzipIdent, src[0:3])
+}
+
+// Uncompress returns gzip uncompressed.
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go:21]: Uncompress returns uncompressed gzip, or error if input is not gzipped or too large.

File: chain/pkg/gzip/gzip.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+	"io/ioutil"
+)
+
+// magic bytes to identify gzip.
+// See https://www.ietf.org/rfc/rfc1952.txt
+// and https://github.com/golang/go/blob/master/src/net/http/sniff.go#L186
+var gzipIdent = []byte("\x1F\x8B\x08")
+
+// IsGzip returns true if the input file is gzipped file.
+func IsGzipped(src []byte) bool {
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go:17]: if the input is gzipped. (remove file)

File: chain/x/oracle/handler.go [hunk 13]
```diff
@@ -37,6 +38,17 @@ func NewHandler(k Keeper) sdk.Handler {
 }
 
 func handleMsgCreateDataSource(ctx sdk.Context, k Keeper, m MsgCreateDataSource) (*sdk.Result, error) {
+	if gzip.IsGzipped(m.Executable) {
+		executable, err := gzip.Uncompress(m.Executable, types.MaxExecutableSize)
```
Comment (sorawit) [inline chain/x/oracle/handler.go]: Declare a new error code

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/pkg/gzip/gzip.go (hunk 10 above)
Comment (sorawit) [inline chain/pkg/gzip/gzip.go:35]: fmt.Errorf("Uncompressed file exceeds max size of %d", maxSize)

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

File: chain/pkg/gzip/gzip.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+	"io"
+	"io/ioutil"
+)
+
+// magic bytes to identify gzip.
+// See https://www.ietf.org/rfc/rfc1952.txt
+// and https://github.com/golang/go/blob/master/src/net/http/sniff.go#L186
+var gzipIdent = []byte("\x1F\x8B\x08")
+
+// IsGzip returns true if the input is gzipped. (remove file)
```
Comment (taobun) [inline chain/pkg/gzip/gzip.go]: IsGzipped

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — DISMISSED** [review]

File: chain/x/oracle/handler_test.go [hunk 15]
```diff
@@ -21,6 +23,21 @@ func mockDataSource(ctx sdk.Context, keeper Keeper) (*sdk.Result, error) {
 	return handleMsgCreateDataSource(ctx, keeper, msg)
 }
 
+func mockDataSourceWithGzipExcuteble(ctx sdk.Context, keeper Keeper) (*sdk.Result, error) {
```
Comment (sorawit) [inline chain/x/oracle/handler_test.go]: Typo

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by taobun — DISMISSED** [review]

File: chain/pkg/gzip/gzip.go [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+	"io"
+	"io/ioutil"
+)
+
+// magic bytes to identify gzip.
+// See https://www.ietf.org/rfc/rfc1952.txt
+// and https://github.com/golang/go/blob/master/src/net/http/sniff.go#L186
+var gzipIdent = []byte("\x1F\x8B\x08")
+
+// IsGzipped returns true if the input is gzipped. (remove file)
```
Comment (sorawit) [inline chain/pkg/gzip/gzip.go]: Please fix comment

File: chain/x/oracle/handler.go [hunk 17]
```diff
@@ -37,9 +38,15 @@ func NewHandler(k Keeper) sdk.Handler {
 }
 
 func handleMsgCreateDataSource(ctx sdk.Context, k Keeper, m MsgCreateDataSource) (*sdk.Result, error) {
-	filename := k.AddFile(m.Executable)
+	var err error
+	if gzip.IsGzipped(m.Executable) {
+		m.Executable, err = gzip.Uncompress(m.Executable, types.MaxExecutableSize)
+		if err != nil {
+			return nil, err
```
Comment (sorawit) [inline chain/x/oracle/handler.go]: Please wrap error here and all places below.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/handler.go [hunk 18]
```diff
@@ -37,9 +38,15 @@ func NewHandler(k Keeper) sdk.Handler {
 }
 
 func handleMsgCreateDataSource(ctx sdk.Context, k Keeper, m MsgCreateDataSource) (*sdk.Result, error) {
-	filename := k.AddFile(m.Executable)
+	if gzip.IsGzipped(m.Executable) {
+		executable, err := gzip.Uncompress(m.Executable, types.MaxExecutableSize)
+		if err != nil {
+			return nil, sdkerrors.Wrapf(types.ErrUncompressionFailed, "file: %x", m.Executable)
```
Comment (sorawit) [inline chain/x/oracle/handler.go]: Don't show the whole executable. Just show `err.Error()`

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**